lib/trino-filesystem/src/main/java/io/trino/filesystem/cache/CacheInputFile.java

```diff
@@ -64,7 +67,14 @@ public TrinoInputStream newStream()
     public long length()
             throws IOException
     {
-        return delegate.length();
+        if (length.isEmpty()) {
+            Optional<String> key = keyProvider.getCacheKey(delegate);
+            if (key.isPresent()) {
+                length = OptionalLong.of(cache.cacheLength(delegate, key.orElseThrow()));
```

Comment:
Why cache length when key is absent? Also this looks like double caching. This at least deserves a comment

Only `IcebergCacheKeyProvider` doesn't attach modified time to key. `DeltaLakeCacheKeyProvider` does it.

So potentially your `length` might be wrong.

I know it's convention in `TrinoInputFile` implementations to cache length, but in context of `CacheInputFile` it "feels" wrong.

Reply:
> Only IcebergCacheKeyProvider doesn't attach modified time to key. DeltaLakeCacheKeyProvider does it
> So potentially your length might be wrong.

DeltaLakeCacheKeyProvider does not attach modified time, I think you are confusing it with DefaultCacheKeyProvider which is used in hive. Files are supposed to be immutable in delta and iceberg where this cache is intended to be used. It won't be used in hive.

> I know it's convention in TrinoInputFile implementations to cache length, but in context of CacheInputFile it "feels" wrong.

It's not just convention. There are explicit tests in AbstractTestTrinoFileSystem which expect that when a TrinoInputFile is created with a provided length, then that length must be returned in subsequent assertions even if the file doesn't actually exist. It's the contract of `io.trino.filesystem.TrinoFileSystem#newInputFile(io.trino.filesystem.Location, long)`
We can't delegate to the cache as it's not aware about this pre-declared length.
We can't delegate to actual FS because there is no way to know whether it has a pre-declared length or will make remote call.